Sources/SourceKitLSP/Swift/SemanticTokensManager.swift

```diff
@@ -0,0 +1,90 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+
+/// Keeps track of the semantic tokens that sourcekitd has sent us for given
+/// document snapshots.
+actor SemanticTokensManager {
+  private var semanticTokens: [DocumentSnapshot.ID: [SyntaxHighlightingToken]] = [:]
```

Comment:
I would prefer to keep it as-is. If we ever decide that we need to store semantic tokens for older version as well, it’s easier to update and conceptually we are storing semantic tokens a document snapshot, the cache eviction logic is separate from that.

Reply:
> conceptually we are storing semantic tokens a document snapshot

Heh, I see this as the other way. We're storing tokens for the latest version of a document. The request for tokens is really "give me the latest tokens of the document which should match version <n>". If it doesn't, then we're out of date and the response isn't likely to matter anyway.

Or to put another away:

> If we ever decide that we need to store semantic tokens for older version as well

Doesn't seem like a use case we care about.

> We still need to keep track of the document contents for clang so that we can re-open the documents in clangd if clangd crashes (for which we need to know the current document contents).

🙇